default.nix

```diff
@@ -118,30 +125,46 @@ with stdenv.lib; let
 in rec {
   mkNodeModules = { src, extraEnvVars ? {} }:
     let
+      localDeps = map dirOfLocal (builtins.filter isLocal (builtins.attrValues (lock.dependencies)));
+
+      origSrc = if src ? _isLibCleanSourceWith then src.origSrc else src;
+      getRelativePath = path: removePrefix (toString origSrc + "/") path;
+
       # filter out everything except package.json and package-lock.json if possible
       # allows to avoid rebuilding node_modules if these two files didn't change
       filteredSrc =
         let
-          origSrc = if src ? _isLibCleanSourceWith then src.origSrc else src;
-          getRelativePath = path: removePrefix (toString origSrc + "/") path;
-          usedPaths = [ "package.json" "package-lock.json" ];
+          usedPaths = [ "package.json" "package-lock.json" ]
+            ++ localDeps;
+          dirOfRec = x: if dirOf x == "." || dirOf x == "/" then [x] else ([x] ++ dirOfRec (dirOf x));
+          usedPathsRec = builtins.concatMap dirOfRec usedPaths;
           cleanedSource = cleanSourceWith {
             src = src;
-            filter = path: type: elem (getRelativePath path) usedPaths;
+            filter = path: type: any (x: elem x usedPathsRec) (dirOfRec (getRelativePath path));
```

Comment:
I just noticed this, but I think `any` should not be used here, because it will include all parent directories of a dependency, while we only want the exact directory of the dependency to be included. For example, if we have dependency `file:./deps/foo`, and there also exists `./deps/extra-stuff` directory which is not used as a dependency, it will still be included in the filtered source